blocks/bricks/recurrent.py

```diff
@@ -630,3 +631,140 @@ def apply(self, *args, **kwargs):
+
+    """
+    def __init__(self, transitions, prototype=None, **kwargs):
+        super(RecurrentStack, self).__init__(**kwargs)
+
+        for d, transition in enumerate(transitions):
+            transition.name += '_%d' % d
+        self.transitions = transitions
+
+        if prototype is None:
+            # By default use Linear (with bias) as a default fork.
+            # This overrides the bad default inside Fork which is without bias.
+            # Yes I know use_bias=True us the defaut, but it is a big deal so
+            # I write it down explicitly.
+            prototype = Linear(use_bias=True)
+        depth = len(transitions)
+        self.forks = [Fork(self.normal_inputs(d), prototype=prototype)
+                      for d in range(1, depth)]
+
+        self.children = self.transitions + self.forks
+
+        # Programmatically fix the parameters of the  recurrent decoration
+        setattr(self.apply, 'sequences', transitions[0].apply.sequences)
+        for t in ["states", "contexts", "outputs"]:
+            v = ['%s_%d' % (s, d)
+                 for d, transition in enumerate(transitions)
+                 for s in getattr(transition.apply, t)]
+            setattr(self.apply, t, v)
+
+    def normal_inputs(self, d):
+        transition = self.transitions[d]
+        return [name for name in transition.apply.sequences
+                if 'mask' not in name]
+
+    def _push_allocation_config(self):
+        # Configure the forks that connect the "states" element in the `states`
+        # of one layer to the elements in the `sequences` of the next layer,
+        # excluding "mask".
+        # This involves `get_dim` requests
+        # to the transitions. To make sure that it answers
+        # correctly we should finish its configuration first.
+        for transition in self.transitions:
+            transition.push_allocation_config()
+
+        for d, fork in enumerate(self.forks):
+            fork.input_dim = self.transitions[d].get_dim("states")
+            fork.output_dims = self.transitions[d+1].get_dims(
+                fork.output_names)
+
+    @recurrent(sequences=[], states=[], outputs=[], contexts=[])
+    def apply(self, **kwargs):
+        """Apply the stack of transitions.
+
+        Parameters
+        ----------
+        All parameters are of type :class:`~tensor.TensorVariable`.
+
+        Returns
+        -------
+        All return values are of type :class:`~tensor.TensorVariable`.
+
+        """
+        results = []
+        for d, transition in enumerate(self.transitions):
+            sequences_names = transition.apply.sequences
+            if d == 0:
+                layer_kwargs = dict(
+                    (s, kwargs.get(s)) for s in sequences_names)
+            else:
+                layer_kwargs = dict(zip(
+                    self.normal_inputs(d),
+                    self.forks[d-1].apply(kwargs["states_%d" % (d-1)],
+                                          as_list=True)))
```

Comment:
In this line I am using the states from the previous step in the sequence as the input to all of the layers (sync mode)
there is also the async option in which the output state of one layer is used as the input (through fork) of the next layer for the same step (lets call this async mode.) which one should I use?